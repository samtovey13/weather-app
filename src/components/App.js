import '../styles/App.css';
import React, { useEffect, useState } from 'react';
import axios from 'axios';
import LocationDetails from './location-details';
import ForecastSummaries from './forecast-summaries';
import ForecastDetails from './forecast-details';
import SearchForm from './search-form';

const App = () => {
  const [forecasts, setForecasts] = useState([]);
  const [location, setLocation] = useState({ city: 'Edinburgh', country: 'GB' });
  const [selectedDate, setSelectedDate] = useState(0);
  const [errorMessage, setErrorMessage] = useState("");

  const getForecasts = async (city) => {
    await axios.get(`https://mcr-codes-weather.herokuapp.com/forecast?city=${city}`)
    .then(res => {
      if (res.status === 200) {
        setLocation(res.data.location);
        setForecasts(res.data.forecasts);
        setErrorMessage("");
      }
    })
    .catch((error) => {
      if (error.response.status === 404) {
        setErrorMessage(`Sorry, ${city} isn't available. Please choose another city.`)
      } else if (error.response.status === 500) {
        setErrorMessage("Oops! Something went wrong. Please try again later.")
      }
    })
  }
  
  useEffect( () => {
    getForecasts(location.city)
  }, [location.city]);

  const selectedForecast = forecasts.find(forecast => 
    forecast.date === selectedDate);
  
  const handleForecastSelect = (date) => {
    setSelectedDate(date);
  }

  const handleSearchText = (searchText) => {
    getForecasts(searchText);
  };
  

  return <div className="forecast">
    <div className="location-details-wrapper">
      <LocationDetails 
        city={location.city} 
        country={location.country}
      />
      <SearchForm
        handleSearchText={handleSearchText}
        errorMessage={errorMessage}
      />
    </div>
    <ForecastSummaries 
      forecasts={forecasts}
      onForecastSelect={handleForecastSelect}
    />
    {
      selectedForecast && (<ForecastDetails forecast={selectedForecast} /> )
    }
    
  </div>
};

export default App;
