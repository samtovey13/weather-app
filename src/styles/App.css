.forecast {
  margin: 0 20px;
  font-family: Arial, Helvetica, sans-serif;
}