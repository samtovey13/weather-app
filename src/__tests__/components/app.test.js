import React from "react";
import { render, waitFor } from "@testing-library/react";
import App from '../../components/App';

import getForecastData from '../../services/getForecasts';
jest.mock('../../services/getForecasts');

describe("App", () => {

  beforeEach(() => {
    getForecastData.mockResolvedValue({
      status: 200,
      data: {
        location: { city: "Manchester", country: "GB" },
        forecasts: [
          {
            date: 1525046400000,
            temperature: {
              max: 11,
              min: 4,
            },
            wind: {
              speed: 13,
              direction: "s",
            },
            humidity: 30,
            description: "Clear",
            icon: "800",
          },
        ],
      },
    });
  });
  

  it('renders the loading screen correctly', async () => {
    getForecastData.mockReturnValue(new Promise(() => {}));
    const { asFragment, getByText } = render(<App />);
    expect(getByText(/loading/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly after loading forecasts', async () => {
    const { asFragment, getByText, queryByText } = render(<App />);
    await waitFor(() => expect(getByText("Clear")).toBeInTheDocument());
    expect(asFragment()).toMatchSnapshot();
    expect(queryByText(/loading/i)).not.toBeInTheDocument();
  });

  it('renders a 404 error correctly', async () => {
    getForecastData.mockResolvedValue({response: {status: 404}});
    const { asFragment, getByText } = render(<App />);
    await waitFor(() => expect(getByText(/please choose another city/i)).toBeInTheDocument());
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders any other errors correctly', async () => {
    getForecastData.mockResolvedValue(Promise.reject("FAILED"));
    const { asFragment, getByText } = render(<App />);
    await waitFor(() => expect(getByText(/something went wrong/i)).toBeInTheDocument());
    expect(asFragment()).toMatchSnapshot();
  });

})
