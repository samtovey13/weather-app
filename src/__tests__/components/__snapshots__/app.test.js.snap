// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders a 404 error correctly 1`] = `
<DocumentFragment>
  <div
    class="forecast"
  >
    <div
      class="location-details-wrapper"
    >
      <h1
        class="location-details"
      >
        Edinburgh, GB
      </h1>
      <div
        class="search-form"
      >
        <div
          id="search-input-wrapper"
        >
          <input
            class="search-input"
            id="location-search-input"
            placeholder="Find a city"
            type="search"
          />
          <button
            class="search-button"
          >
            <i
              class="fa fa-search"
            />
          </button>
        </div>
        <div
          class="error-message"
        >
          Sorry, Edinburgh isn't available. Please choose another city.
        </div>
      </div>
    </div>
    <div
      class="forecast-summaries"
    />
  </div>
</DocumentFragment>
`;

exports[`App renders any other errors correctly 1`] = `
<DocumentFragment>
  <div
    class="forecast"
  >
    <div
      class="location-details-wrapper"
    >
      <h1
        class="location-details"
      >
        Edinburgh, GB
      </h1>
      <div
        class="search-form"
      >
        <div
          id="search-input-wrapper"
        >
          <input
            class="search-input"
            id="location-search-input"
            placeholder="Find a city"
            type="search"
          />
          <button
            class="search-button"
          >
            <i
              class="fa fa-search"
            />
          </button>
        </div>
        <div
          class="error-message"
        >
          Oops! Something went wrong. Please try again later.
        </div>
      </div>
    </div>
    <div
      class="forecast-summaries"
    />
  </div>
</DocumentFragment>
`;

exports[`App renders correctly after loading forecasts 1`] = `
<DocumentFragment>
  <div
    class="forecast"
  >
    <div
      class="location-details-wrapper"
    >
      <h1
        class="location-details"
      >
        Manchester, GB
      </h1>
      <div
        class="search-form"
      >
        <div
          id="search-input-wrapper"
        >
          <input
            class="search-input"
            id="location-search-input"
            placeholder="Find a city"
            type="search"
          />
          <button
            class="search-button"
          >
            <i
              class="fa fa-search"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      class="forecast-summaries"
    >
      <div
        class="forecast-summary"
      >
        <span
          class="date"
          data-testid="date-id"
        >
          Mon 30th Apr
        </span>
        <span
          class="temperature"
          data-testid="temperature-id"
        >
          11Â°c
        </span>
        <span
          class="description"
          data-testid="description-id"
        >
          Clear
        </span>
        <span
          class="icon"
          data-testid="icon-id"
        >
          <i
            class="wi wi-day-sunny"
            data-testid="weather-icon-id"
            name="owm"
          />
        </span>
        <button
          class="details-button"
        >
          More Details
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`App renders the loading screen correctly 1`] = `
<DocumentFragment>
  <div
    class="loading-message-wrapper"
  >
    <div
      class="loading-message"
    >
      ...loading...
    </div>
  </div>
</DocumentFragment>
`;
